plugins/rbac-backend/src/service/permission-policy.ts

```diff
@@ -61,61 +101,163 @@ const useAdmins = async (admins: Config[], enf: Enforcer) => {
     'allow',
   ];
   if (!(await enf.hasPolicy(...adminUpdatePermission))) {
-    await enf.addPolicy(...adminUpdatePermission);
+    await enf.addPolicy(adminUpdatePermission, 'configuration');
   }
 };
 
 const addPredefinedPoliciesAndGroupPolicies = async (
   preDefinedPoliciesFile: string,
-  enf: Enforcer,
+  enf: EnforcerDelegate,
+  roleMetadataStorage: RoleMetadataStorage,
+  knex: Knex,
 ) => {
   const fileEnf = await newEnforcer(
     newModelFromString(MODEL),
     new FileAdapter(preDefinedPoliciesFile),
   );
-  const policies = await fileEnf.getPolicy();
+  const policies = new Set<string[]>(await fileEnf.getPolicy());
+  const groupPolicies = new Set<string[]>(await fileEnf.getGroupingPolicy());
+
+  const oldFilePolicies = new Set<string[]>();
+  const policiesMetadata = await enf.getFilteredPolicyMetadata('csv-file');
+  for (const policyMetadata of policiesMetadata) {
+    oldFilePolicies.add(metadataStringToPolicy(policyMetadata.policy));
+  }
+
+  const policiesToDelete: string[][] = [];
+  const groupPoliciesToDelete: string[][] = [];
+  for (const oldFilePolicy of oldFilePolicies) {
+    if (oldFilePolicy.length === 2 && !groupPolicies.has(oldFilePolicy)) {
+      groupPoliciesToDelete.push(oldFilePolicy);
+    } else if (!policies.has(oldFilePolicy)) {
+      policiesToDelete.push(oldFilePolicy);
+    }
+  }
+
+  const rolesToDelete = new Set<string>();
+  groupPoliciesToDelete
+    .filter(gp => gp[1].startsWith(`role:`))
+    .forEach(gp => {
+      rolesToDelete.add(gp[1]);
+    });
+  for (const gp of groupPolicies) {
+    if (rolesToDelete.has(gp[1])) {
+      rolesToDelete.delete(gp[1]);
+    }
+  }
+
+  const delRoleMetaTrx = await knex.transaction();
+  try {
+    for (const roleMeta of rolesToDelete) {
+      await roleMetadataStorage.removeRoleMetadata(roleMeta, delRoleMetaTrx);
+    }
+    await enf.removeGroupingPolicies(groupPoliciesToDelete, true);
+    await delRoleMetaTrx.commit();
+  } catch (err) {
+    await delRoleMetaTrx.rollback();
+    throw err;
+  }
```

Comment:
Could we possibly refactor this to simplify this and maybe fix the issue with using `better-sqlite3`. Something along the lines of moving this over to the `EnforerDelegate` under `removeGroupingPolicies`

```
 async removeGroupingPolicies(
    policies: string[][],
    allowToDeleCSVFilePolicy?: boolean,
  ): Promise<void> {
    const rmMetadataTrx = await this.knex.transaction();
    try {
      for (const policy of policies) {
        await this.checkIfPolicyModifiable(policy, allowToDeleCSVFilePolicy);
        await this.metadataStorage.removePolicyMetadata(policy, rmMetadataTrx);
        
        // We would need to manipulate the policy to get just the role information
        await this.roleMetadataStorage.removeRoleMetadata(policy, rmMetadataTrx);

      }
      const ok = await this.enforcer.removeGroupingPolicies(policies);
      if (!ok) {
        throw new Error(
          `Failed to delete grouping policies: ${policiesToString(policies)}`,
        );
      }
      await rmMetadataTrx.commit();
    } catch (err) {
      await rmMetadataTrx.rollback(err);
      throw err;
    }
  }
```
Hopefully, this would reduce the amount of transactions that we would need to create. Also, we could do this for all areas where we are working with role metadata. What do you think?

Reply:
I merged your pull request, so I think we handled this proposal.